src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -186,118 +171,157 @@ public bool Contains(TKey key)
         {
-            int numItemsRemoved = 0;
+#pragma warning disable CA1031 // Do not catch general exception types
             try
             {
-                var node = _doubleLinkedList.First;
+                LinkedListNode<LRUCacheItem<TKey, TValue>> node = _doubleLinkedList.First;
                 while (node != null)
                 {
-                    var nextNode = node.Next;
+                    LinkedListNode<LRUCacheItem<TKey, TValue>> nextNode = node.Next;
                     if (node.Value.ExpirationTime < DateTime.UtcNow)
                     {
                         _doubleLinkedList.Remove(node);
-                        if (_map.TryRemove(node.Value.Key, out var cacheItem))
-                            OnItemRemoved?.Invoke(cacheItem.Value);
-
-                        numItemsRemoved++;
+                        if (_map.TryRemove(node.Value.Key, out LRUCacheItem<TKey, TValue> cacheItem))
+                            OnItemExpired?.Invoke(cacheItem.Value);
                     }
 
                     node = nextNode;
                 }
             }
-            catch (ObjectDisposedException ex)
+            catch(Exception ex)
             {
                 if (LogHelper.IsEnabled(EventLogLevel.Warning))
                     LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10902, LogHelper.MarkAsNonPII(nameof(RemoveExpiredValuesLRU)), ex));
             }
-
-            return numItemsRemoved;
+            finally
+            {
+                _removeExpiredValuesState = ActionNotQueued;
+                _timeForNextExpiredValuesRemoval = DateTime.UtcNow.AddSeconds(_removeExpiredValuesIntervalInSeconds);
+            }
+#pragma warning restore CA1031 // Do not catch general exception types
         }
 
         /// <summary>
         /// Remove all expired cache items from the _map ONLY. This is called for the non-LRU (_maintainLRU = false) scenaro.
         /// The enumerator returned from the dictionary is safe to use concurrently with reads and writes to the dictionary, according to the MS document.
         /// </summary>
         /// <returns>Number of items removed.</returns>
-        internal int RemoveExpiredValues()
+        internal void RemoveExpiredValues()
         {
-            int numItemsRemoved = 0;
+#pragma warning disable CA1031 // Do not catch general exception types
             try
             {
-                foreach (var node in _map)
+                foreach (KeyValuePair<TKey, LRUCacheItem<TKey, TValue>> node in _map)
                 {
                     if (node.Value.ExpirationTime < DateTime.UtcNow)
                     {
                         if (_map.TryRemove(node.Value.Key, out var cacheItem))
-                            OnItemRemoved?.Invoke(cacheItem.Value);
+                            OnItemExpired?.Invoke(cacheItem.Value);
+                    }
+                }
+            }
+            catch
+            {
```

Comment:
why not log here?

Reply:
@keegan-caruso didn't see a reason as there could be many, many removals.

Reply:
I added it, we may be able to use it.